src/shims/mod.rs

```diff
@@ -43,11 +43,43 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         // Try to see if we can do something about foreign items.
         if this.tcx.is_foreign_item(instance.def_id()) {
-            // An external function that we cannot find MIR for, but we can still run enough
-            // of them to make miri viable.
-            this.emulate_foreign_item(instance.def_id(), args, dest, ret)?;
-            // `goto_block` already handled.
-            return Ok(None);
+            let attrs = this.tcx.get_attrs(instance.def_id());
+            let link_name = match attr::first_attr_value_str_by_name(&attrs, sym::link_name) {
+                Some(name) => name.as_str(),
+                None => this.tcx.item_name(instance.def_id()).as_str(),
+            };
+            // Strip linker suffixes (seen on 32-bit macOS).
+            let link_name = link_name.trim_end_matches("$UNIX2003");
+
+            // We handle "panic_impl" differently from other foreign items,
+            // since we actually want to load existing MIR for it.
+            if this.tcx.lang_items().panic_impl() == Some(instance.def.def_id()) {
+                // Manually forward to 'panic_impl' lang item
+                let panic_impl_real = this.tcx.lang_items().panic_impl().unwrap();
+                return Ok(Some(this.load_mir(InstanceDef::Item(panic_impl_real), None)?));
+            } else {
+                // An external function that we (possibly) cannot find MIR for, but we can still run enough
+                // of them to make miri viable.
+                this.emulate_foreign_item(link_name, args, dest, ret, unwind)?;
+                return Ok(None)
+            }
+        }
+
+        let is_panic_fn = this.tcx.lang_items().panic_fn() == Some(instance.def.def_id());
+        let is_begin_panic = this.tcx.lang_items().begin_panic_fn() == Some(instance.def.def_id());
+
+        // Run the common hooks also supported by CTFE.
+        // We do this after our custom handling logic above,
+        // so that we can override the default behavior if necessary
+        // Note that we deliberately do *not* invoke the common hooks
+        // for 'begin_panic_fn' and 'is_panic_fn'. We want to invoke the actual MIR in libstd,
+        // so that we enter the libstd pre-panic code (which will eventually
+        // invoke '__rust_start_panic'
```

Comment:
Can we change the organization in librustc_mir so that there is a function we can call that only hooks what Miri really wants hooked (and not the panic stuff)? Then we could avoid this hack here.

Reply:
Actually `hook_fn`, right now, hooks panics and nothing else. So we can just drop the call here (and in follow-up cleanup, rename it to `hook_panic_fn` or so).